### PR TITLE
Upload validation

### PR DESCRIPTION
**Pull request by jorisvaesen**
Started porting upload validation from the 2.x plugin.
It's a standalone validation provider which can optionally be loaded and used.

I've split it up into multiple traits to make it more manageable and give the option to add specific validations (video, pdf, ...) later.

Usage:
`$validator->provider('upload', \Josegonzalez\Upload\Validation\UploadValidation::class);`

If you only want to use the image validation rules:
`$validator->provider('upload', \Josegonzalez\Upload\Validation\ImageValidation::class);`

After this you can use it as:
`$validator->add('file', 'filePhpUploadSize', [
            'rule' => 'isUnderPhpSizeLimit',
            'message' => 'File is too large to upload',
            'provider' => 'upload'
]);`

If this code could makes it, i'll write some documentation for it.

### CODE REVIEW

**Comment by josegonzalez** [comment]
Needs 100% test coverage, but I'm on board with the functionality.

**Comment by josegonzalez** [comment]
Getting there!

File: docs/validation.rst [hunk 1]
```diff
@@ -0,0 +1,173 @@
...
+        // OR
+        $validator->provider('upload', \Josegonzalez\Upload\Validation\DefaultValidation::class);
+        
+    ?>
+    
+Afterwards, you can use its rules like:
+
+.. code:: php
+
+    <?php
```
Comment (josegonzalez) [inline docs/validation.rst]: Do we need the php open/close tags?

File: docs/validation.rst [hunk 2]
```diff
@@ -0,0 +1,173 @@
...
+    
+    ?>
+    
+
+UploadValidation
+^^^^^^^^^^^^^^^^
+
+**isUnderPhpSizeLimit**
+
+Check that the file does not exceed the max file size specified by PHP
```
Comment (josegonzalez) [inline docs/validation.rst]: Can you add an example for *each* one?

File: src/Validation/Traits/ImageValidationTrait.php [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+     *
+     * @param mixed $check Value to check
+     * @param int $width Width of Image
+     * @return bool Success
+     */
+    public static function isAboveMinWidth($check, $width)
+    {
+        // Non-file uploads also mean the height is too big
+        if (!isset($check['tmp_name']) || !strlen($check['tmp_name'])) {
+            return false;
```
Comment (josegonzalez) [inline src/Validation/Traits/ImageValidationTrait.php]: Is there a way to return true if there wasn't an image in the first place? The 2.0 version had an option to enable these checks but disable them *if* there wasn't an upload.

File: docs/validation.rst (hunk 1 above)
Comment (jorisvaesen) [inline docs/validation.rst]: This is how its done in 2.x: https://github.com/josegonzalez/cakephp-upload/blob/2.x/docs/validation.rst

File: src/Validation/Traits/ImageValidationTrait.php (hunk 3 above)
Comment (jorisvaesen) [inline src/Validation/Traits/ImageValidationTrait.php]: This rule should not be used alone but more in combination with isSuccessfulWrite and mimeType.

Comment (josegonzalez) [inline src/Validation/Traits/ImageValidationTrait.php]: So we don't allow conditional validation?

Comment (jorisvaesen) [inline src/Validation/Traits/ImageValidationTrait.php]: Conditional validation can be done through a callback on the 'on' parameter: http://book.cakephp.org/3.0/en/core-libraries/validation.html#conditional-validation

Comment (josegonzalez) [inline src/Validation/Traits/ImageValidationTrait.php]: Please do.

**Comment by josegonzalez** [comment]
Looks good, ship it.

**Pull request merged**